[single-shot] Completion for new single-shot API

kparichay opened the pull request:
Completed the implementation for single-shot API without GStreamer
Added a separate binary for this API
Added remaining functions in `tensor_filter_single.c`
Added another test run in meson and `nnstreamer.spec` for testing of this new implementation

Note: single-shot new API is not part of the original `capi-nnstreamer-0.2.1-0.*.rpm` package
It is contained in `capi-nnstreamer-single-new-0.2.1-0.*.rpm` along with its devel version

Working on #1547

Self evaluation:
Build test: [x]Passed [ ]Failed [ ]Skipped
Run test: [x]Passed [ ]Failed [ ]Skipped

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com

jaeyun-jung commented on api/capi/include/nnstreamer.h:
```diff
@@ -707,6 +707,18 @@ int ml_tensors_info_get_tensor_dimension (ml_tensors_info_h info, unsigned int i
+ * @retval #ML_ERROR_NOT_SUPPORTED Not supported.
+ * @retval #ML_ERROR_INVALID_PARAMETER Given parameter is invalid.
+ */
+int ml_tensors_data_create_no_alloc (const ml_tensors_info_h info, ml_tensors_data_h *data);
```
I think this is internal function. Plz move the declaration into -private.h.

kparichay replied:
Thanks for the comment, updated.

kparichay commented:
V2:
Moved `ml_tensors_data_create_no_alloc` from `api/capi/nnstreamer.h` to `api/capi/nnstreamer-capi-private.h`

jaeyun-jung commented on api/capi/src/nnstreamer-capi-util.c:
```diff
@@ -529,6 +530,29 @@ ml_tensors_data_create (const ml_tensors_info_h info,
+  ml_tensors_data_s *_data = NULL;
+  gint i;
+
+	status = ml_tensors_data_create_no_alloc (info, (ml_tensors_data_h *) &_data);
```
plz fix indent in this function.

kparichay replied:
done

leemgs commented:
@kparichay , Could you archive the changes by adding V1, V2, V3,..... ?
* FYI : https://github.com/nnsuite/nnstreamer-example/pull/26

kparichay commented:
V3:
Updated indentation in `nnstreamer-capi-util.c`
Rebased

V2:
Moved `ml_tensors_data_create_no_alloc` from `api/capi/nnstreamer.h` to `api/capi/nnstreamer-capi-private.h`

jaeyun-jung commented on api/capi/src/nnstreamer-capi-single-new.c:
```diff
@@ -402,13 +407,83 @@ int
+    return ML_ERROR_PERMISSION_DENIED;
+
+  /** TODO: create a new thread, which will invoke and wait with a timeout */
+  status = klass->invoke (single_h->filter, in_tensors, out_tensors);
```
invoke() returns boolean.

kparichay replied:
updated.

jaeyun-jung commented on api/capi/src/tensor_filter_single.c:
```diff
@@ -374,12 +573,42 @@ static gboolean
+    }
+  }
+
+  status = self->fw->invoke_NN (&self->prop, &self->privateData, input, output);
```
check return type (int maybe)

kparichay replied:
updated.

kparichay commented:
V4:
appropriate handling of return from `invoke` and `invoke_NN`

V3:
Updated indentation in `nnstreamer-capi-util.c`
Rebased

V2:
Moved `ml_tensors_data_create_no_alloc` from `api/capi/nnstreamer.h` to `api/capi/nnstreamer-capi-private.h`

myungjoo commented on api/capi/src/nnstreamer-capi-single-new.c:
```diff
@@ -402,13 +407,82 @@ int
+  for (i = 0; i < in_data->num_tensors; i++) {
+    size_t raw_size = ml_tensor_info_get_size (&single_h->in_info.info[i]);
+
+    if (!in_data->tensors[i].tensor || in_data->tensors[i].size != raw_size) {
```
What if the model may accept different input sizes and the API user knows it and wants to enter various input dimensions?

kparichay replied:
Do you mean that user wants to set different input dimension for each invoke? if thats the case, a new function can be added to update input and output info, which can be called before invoke (in a different PR I think).

myungjoo replied:
There are models that have flexible input dimensions. The current version of tensorflow-lite doesn't, though. E.g., many of our custom-filter examples have flexible input dimensions as well.

myungjoo replied:
Anyway.. this is not for 5.5 M2, but for 6.0+; thus we may discuss updating (preferably adding more semantics to it) the API.

myungjoo replied:
Please add an issue item for "Flexible Input Dimension for Single-Shot API". I think we may address this issue later independently.

kparichay replied:
Added #1651

dongju-chae commented on api/capi/src/nnstreamer-capi-single-new.c:
```diff
@@ -402,13 +407,82 @@ int
+  *output = NULL;
+
+  if (!single_h->filter) {
+    return ML_ERROR_INVALID_PARAMETER;
```
How about adding an error msg like other cases?

kparichay replied:
Done.

dongju-chae commented on api/capi/src/tensor_filter_single.c:
```diff
@@ -374,12 +573,40 @@ static gboolean
+  if (self->fw->allocate_in_invoke == FALSE)
+    for (i = 0; i < self->prop.output_meta.num_tensors; i++)
+      g_free (output[i].data);
+  return FALSE;
```
What if `g_malloc()` is failed in the middle of index? I think i should be the latest index which is successful.

kparichay replied:
I am not sure if I understand correctly. If you are concerned about calling `g_free()` on index `i` which have not been called with `g_malloc`, `g_free` can work with `NULL` as well.
Note: If `g_malloc` is not called, `output[i].data` is NULL.

dongju-chae replied:
ohh then ok
